primitives/src/signatures/bls.rs

```diff
@@ -1,157 +1,316 @@
 use crate::{
     constants::{
-        BLS_SIG_KEY_SIZE, BLS_SIG_SIGNATURE_SIZE, BLS_SIG_VERKEY_SIZE, CS_ID_BLS_SIG_NAIVE,
+        BLS_SIG_COMPRESSED_PK_SIZE, BLS_SIG_COMPRESSED_SIGNATURE_SIZE, BLS_SIG_PK_SIZE,
+        BLS_SIG_SIGNATURE_SIZE, BLS_SIG_SK_SIZE, CS_ID_BLS_MIN_SIG,
     },
     errors::PrimitivesError,
 };
 
-use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, SerializationError};
+use ark_serialize::*;
 use ark_std::{
     convert::TryInto,
     format,
+    ops::{Deref, DerefMut},
     rand::{CryptoRng, RngCore},
+    vec::Vec,
 };
-
-use blst::BLST_ERROR;
+use blst::{min_sig::*, BLST_ERROR};
 use espresso_systems_common::jellyfish::tag;
 use tagged_base64::tagged;
+use zeroize::{Zeroize, Zeroizing};
 
-pub use blst::min_sig::{PublicKey, SecretKey, Signature};
-use zeroize::Zeroize;
-
-/// Newtype wrapper for a BLS Signing Key.
 #[tagged(tag::BLS_SIGNING_KEY)]
-#[derive(Clone, Debug, Zeroize)]
+#[derive(Clone, Debug, Default, Zeroize)]
+#[zeroize(drop)]
+/// A BLS Secret Key (Signing Key).
 pub struct BLSSignKey(SecretKey);
 
-impl core::ops::Deref for BLSSignKey {
+impl Deref for BLSSignKey {
     type Target = SecretKey;
-
     fn deref(&self) -> &Self::Target {
         &self.0
     }
 }
 
 impl CanonicalSerialize for BLSSignKey {
-    fn serialized_size(&self) -> usize {
-        BLS_SIG_KEY_SIZE
+    fn serialize<W: Write>(&self, writer: W) -> Result<(), SerializationError> {
+        CanonicalSerialize::serialize(&self.to_bytes()[..], writer)
     }
 
-    fn serialize<W: ark_serialize::Write>(&self, writer: W) -> Result<(), SerializationError> {
-        let bytes = &self.0.serialize();
-        CanonicalSerialize::serialize(bytes.as_ref(), writer)
+    fn serialized_size(&self) -> usize {
+        BLS_SIG_SK_SIZE
     }
 }
 
 impl CanonicalDeserialize for BLSSignKey {
-    fn deserialize<R: ark_serialize::Read>(mut reader: R) -> Result<Self, SerializationError> {
+    fn deserialize<R: Read>(mut reader: R) -> Result<Self, SerializationError> {
         let len = <usize as ark_serialize::CanonicalDeserialize>::deserialize(&mut reader)?;
-        if len != BLS_SIG_KEY_SIZE {
+        if len != BLS_SIG_SK_SIZE {
             return Err(SerializationError::InvalidData);
         }
 
-        let mut key = [0u8; BLS_SIG_KEY_SIZE];
-        reader.read_exact(&mut key)?;
-        SecretKey::deserialize(&key)
+        let mut sk_bytes = [0u8; BLS_SIG_SK_SIZE];
+        reader.read_exact(&mut sk_bytes)?;
+        SecretKey::from_bytes(&sk_bytes)
             .map(Self)
             .map_err(|_| SerializationError::InvalidData)
     }
 }
 
 impl PartialEq for BLSSignKey {
     fn eq(&self, other: &Self) -> bool {
-        self.0.serialize() == other.0.serialize()
+        // constant time comparison
+        let xor_res: Vec<u8> = self
+            .to_bytes()
+            .iter()
+            .zip(other.to_bytes().iter())
+            .map(|(a, b)| a ^ b)
+            .collect();
+        xor_res == [0u8; 32]
```

Comment:
under the hood you're calling the same methods anyway, since `to_bytes == serialize`. 
On top of that you're iterating and XORing values, I think the previous method is more performant, and no less constant time.

Reply:
previous methods is doing comparison of two slices? `PartialEq for [u8]` under the hood use `.all()` on the iterator, which will early return, thus no longer constant time. 
(altho admittedly our library has never put constant-time as a high priority, kind of out-of-whim to do it here. 🤔 )

Reply:
I decide to switch back to your original impl. If we are concerned about constant-time, we should do a more thorough audting in the future.

Reply:
> `PartialEq for [u8]` under the hood use `.all()` on the iterator, which will early return

Ah, I wasn't aware of that. Thanks for pointing it out.

But yes in this case I think it's safer to put a simpler impl in place, than start worrying about const time algorithm for `PartialEq` :)